Add Pinned Tabs

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Commands for pinning and unpinning tabs. Fixes #9326. Closes #9614.

This is based on #9614 with a bug fix and a switch to codicons for pinned tab icons.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
1. Use the tabbar context menu or command palette to pin and unpin tabs. 
   > In the main or bottom areas, they should receive a pin icon.
   If dirty, an editor should receive pin-with-dot icon.
2. Try to close pinned editors. They should not close, but should convert to unpinned editors.
3. Reload the page / restart the application. Pinned state should be restored.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

msujew commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -241,6 +242,16 @@ export namespace CommonCommands {
         category: VIEW_CATEGORY,
         label: 'Toggle Status Bar Visibility'
     });
+    export const PIN_TAB = Command.toDefaultLocalizedCommand({
+        id: 'workbench.action.pinEditor',
+        category: VIEW_CATEGORY,
+        label: 'Pin Tab'
+    });
+    export const UNPIN_TAB = Command.toDefaultLocalizedCommand({
+        id: 'workbench.action.unpinEditor',
+        category: VIEW_CATEGORY,
+        label: 'Unpin Tab'
+    });
```
The eslint rule doesn't really work for `toDefaultLocalizedCommand` (I should fix this perhaps)

`Pin` and `Unpin` are the labels used by VSCode and can both be found in the `nls.metadata.json` file.

colin-grant-work replied:
I did wonder about whether the linting would catch it for the other NLS utilities. I'll update the text.

msujew replied:
I actually just noticed that VSCode is using `Pin editor`/`Unpin editor`. Since we do not only allow pinning on editors, we should maybe introduce new translations here? Or is `Pin editor` accurate enough for the command label?

colin-grant-work replied:
I think editors are going to be the most common targets, in any case, so I'll use '(un)pin editor' for now.

msujew reviewed (commented):
I noticed some minor issues in this feature:

1. Pinning a side bar tab prevents it from closing. Moving this pinned tab to the main area leaves it in an invalid state. The usual close icon is displayed and works as expected. However, the context menu entry for `Pin` is disabled and the one for `Unpin` is enabled.
2. Pinning a focused tab removes the focus from the tab. VSCode refocuses the target tab of the context menu after it closes. This one might not have to do with your PR and is probably more of a general issue of how context menus and tab focus work.

colin-grant-work commented on packages/core/src/browser/shell/tab-bars.ts:
```diff
@@ -162,7 +163,10 @@ export class TabBarRenderer extends TabBar.Renderer {
                 this.renderBadge(data, isInSidePanel)
             ),
-            this.renderCloseIcon(data)
+            h.div({
+                className: 'p-TabBar-tabCloseIcon',
+                onclick: this.handleCloseClickEvent
+            })
```
Thanks. I think the original PR predated your work in that area.

colin-grant-work commented:
@msujew, I think the problem was probably code in `ViewContainer`, and I've adjusted that, and adjusted the code checking enablement of the commands to consistently check the `closeable` field, as well, so that `closeable: true` is always treated as `pinned: false` even if the classname is present.

msujew commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -636,6 +658,16 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
         });
+        registry.registerMenuAction(SHELL_TABBAR_CONTEXT_PIN, {
+            commandId: CommonCommands.PIN_TAB.id,
+            label: 'Pin',
```
```suggestion
            label: nls.localizeByDefault('Pin'),
```

msujew commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -636,6 +658,16 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
+        });
+        registry.registerMenuAction(SHELL_TABBAR_CONTEXT_PIN, {
+            commandId: CommonCommands.UNPIN_TAB.id,
+            label: 'Unpin',
```
```suggestion
            label: nls.localizeByDefault('Unpin'),
```

msujew commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -996,6 +1048,16 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
+            },
+            {
+                command: CommonCommands.UNPIN_TAB.id,
+                keybinding: 'ctrlcmd+k shift+enter',
```
I noticed that pinning does work using the specified keyboard shortcut. However, unpinning does not work for some reason.

colin-grant-work replied:
It turns out this was also a consequence of the 'editor' vs. 'widget' question. An earlier version of the PR had either had two context keys or only `activeWidgetIsPinned` and it was switched to `editor` to agree with VSCode, but the keybindings still referred to `widget`.

colin-grant-work commented:
> Alright, I noticed a minor issue with the keyboard shortcut: When an editor is unfocused (for example by clicking on the tab-bar), using the shortcut to pin works, but using it to unpin does not. It works in every other case that I've tested. Since this is quite minor, I will approve this PR nonetheless, as everything else is working.

I appreciate the thorough testing! I will check on this scenario.

vince-fugnitto reviewed (commented):
I confirmed the following :+1: 

- [x] pinning/unpinning works
- [x] pinning/unpinning tabs in the main-area properly decorate the tab with the pin icon
- [x] pinning a editor-preview works, the editor becomes persistent
- [x] pinning dirty editors works, the tab is decorated with the dirty pin icon 
- [x] pinning tabs in the sidepanels work well
- [x] pinned tabs are uncloseable unless unpinned first
- [x] pinned tabs are persistent on reloads
- [x] pinning with the keyboard shortcut works well
- [ ] unpinning with the keyboard shortcut does not seem to work for some widgets (ex: reproducible with search-in-workspace) (likely what Mark noticed too)

colin-grant-work commented:
@msujew, @vince-fugnitto, thanks again for the thorough reviews. I believe I tracked down the problem with unpinning unfocused widgets: the context key setting and the command enablement were looking at different conditions. I have adjusted the code so that the context key is always looking at the same widget as the command will act on.

vince-fugnitto reviewed (approved):
I confirmed that the outstanding issue regarding the keyboard shortcut has been fixed, the last issue in https://github.com/eclipse-theia/theia/pull/10817#pullrequestreview-902060656 now works 👍

msujew reviewed (approved):
Thanks Colin, that last change seems to have done it 👍
